### PR TITLE
mvebu: add support for IIJ SA-W2

### PR DESCRIPTION
**Pull request by musashino205**
This patch series adds simple-fw mtdsplit driver and support for IIJ SA-W2.

### CODE REVIEW

File: target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_simple.c
Comment (BKPepe) [inline target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_simple.c:1]: Any plans to upstream this?

File: target/linux/mvebu/files/arch/arm/boot/dts/armada-380-sa-w2.dts
Comment (BKPepe) [inline target/linux/mvebu/files/arch/arm/boot/dts/armada-380-sa-w2.dts:1]: Shouldn't be this upstreamed?

**Review by BKPepe — CHANGES_REQUESTED** [review]
It would be nice if many things are upstreamed so OpenWrt does not need to carry those files forever.

File: target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_simple.c
Comment (musashino205) [reply target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_simple.c:1]: This is a mtdsplit parser of OpenWrt, for splitting firmware partition to kernel and rootfs.

File: target/linux/mvebu/files/arch/arm/boot/dts/armada-380-sa-w2.dts
Comment (musashino205) [reply target/linux/mvebu/files/arch/arm/boot/dts/armada-380-sa-w2.dts:1]: I don't have any plans to send this for now because I don't have time to maintain dts in upstream...

**Comment by musashino205** [comment]
UPDATE:

- fix network configuration in dts
  - eth0 is used for WAN connection on port4 on stock firmware, not eth1

probably:
(quoted: printed on the case)

```
    +-------------------------+
    |         88F6810         |
    | +----+           +----+ |
    | |eth1|           |eth0| |
    +-++---+-----------+---++-+
       |                   |
       |                   |
+-----++-+---------------+-++-----+
|     |p6|    88E6172    |p5|     |
|     ++-+               +-++     |
|      |                   |      |
|    +-+---+----+----+     +-+    |
|    |     |    |    |       |    |
|   ++-+ +-++ +-++ +-++    +-++   |
|   |p0| |p1| |p2| |p3|    |p4|   |
+---+--+-+--+-+--+-+--+----+--+---+
    "0"  "1"  "2"  "3"     "GE0"
    |                 |
    +-----------------+
           "GE1"
```

**Comment by musashino205** [comment]
UPDATE:

- update flash instruction
  - use ```firmwrite``` command to flash sysupgrade image to spi-nor instead of ```flashwrite```

File: target/linux/mvebu/files/arch/arm/boot/dts/armada-380-sa-w2.dts
Comment (BKPepe) [reply target/linux/mvebu/files/arch/arm/boot/dts/armada-380-sa-w2.dts:1]: So, we should have time here to maintain it for you? :shrug:

File: target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_simple.c
Comment (BKPepe) [reply target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_simple.c:1]: That is possible, but I dont think that we need to have the entirely new file in this repository as there are not enough people, who would maintain it and keep it updated. 

- Are there any plans that other routers will use it as well?
- Why it was necessary to have the simple variant instead of using existing variants?

Comment (musashino205) [reply target/linux/generic/files/drivers/mtd/mtdsplit/mtdsplit_simple.c:1]: > Are there any plans that other routers will use it as well?

I don't know. May or may not use.

> Why it was necessary to have the simple variant instead of using existing variants?

Unless there is a device that I am not aware of, there is no other device that uses the same header, and the existing mtdsplit parsers cannot handle it.
At first, I created a mtdsplit parser exclusively for the header format of this device, but since there was a possibility that this device would be the only model to be used in the future, I changed it to a general-purpose specification.

**Comment by musashino205** [comment]
UPDATE:

- rebase from master

File: target/linux/mvebu/files/arch/arm/boot/dts/armada-380-sa-w2.dts
Comment (tmn505) [reply target/linux/mvebu/files/arch/arm/boot/dts/armada-380-sa-w2.dts:1]: > I don't have any plans to send this for now because I don't have time to maintain dts in upstream...

After dts is accepted there is no need to maintain it upstream, it will be updated by target maintainer or if bindings to driver change, by the sole author of those changes. Unless You forgot something, then after sending changes it's usually accepted without questions.

> In fact,
> 
>     * there are already some device dts files in OpenWrt, I decided sending to upstream is not required

Yes there are and sending upstream is not required, I think.

>     * this dts has some OpenWrt-specific properties/nodes (`led-*` aliases, `firmware` partition, etc...), so I felt that the advantage of sending to upstream at this point is not very big

Yes these are OpenWrt specific, we do patch those in upstream dts and this is fine. Let me give You an example of possible breakage when dts is "out-of-date". At some point, can't remember where was the conversation, kernel bindings of a NAND controller changed and Linksys WRT 32x user reported no rootfs. Guess why only him and not the rest of Linksys WRT series? Because WRT 32x is the only one not upstreamed, the rest of WRT series was fine because author responsible for the changes also adjusted dts. Leaving aside the maintenance burden of changing downstream dts, imagine situation when someone only uses stable releases and that issue slipping through, because no one tested the snapshot. Also imagine possible hurdles recovering from that for casual user. Fortunately that WRT 32x issue was noticed before PR was merged. So please consider upstreaming.

Comment (musashino205) [reply target/linux/mvebu/files/arch/arm/boot/dts/armada-380-sa-w2.dts:1]: Thank you for the great explanation, I was convinced.
OK, I'll try to send dts to upstream...

**Pull request closed without merging**
